[BE][Team04 - Starve, Bibi, YJ, 쏭] Card Class, Repository, CRUD 구현

bibi6666667 opened the pull request:
안녕하세요!
TodoList 미션을 진행하는 04번째 팀 [Starve, Bibi, YJ, 쏭] 입니다.

기본적인 Card Class, Repository, CRUD 구현을 마쳐, 첫 번째 PR을 보냅니다.

## 구현내용
1. BE단 API제공을 위한 기본 구조 (Controller, Domain, Repository, Service) 구현
2. 기본 CRUD 기능 구현
3. 본격적인 SQL과 AWS EC2 사용 전에 h2와 PostMan을 테스트 용도로 사용

## 구현하며 궁금했던 내용
- Card의 상태(todo / doing / done / deleted)를 나타내기 위해 Enum과 boolean 중 어떤 것을 사용할지 고민했습니다.
처음에는 간단히 모든 상태를 변수로 boolean을 사용했는데, 작성하다 보니 Card클래스의 메서드 중복 제거를 위해 Enum을 도입하자는 제안이 나왔습니다. 그런데,
1. 변수가 4개밖에 되지 않는데 굳이 Enum을 사용해 코드의 복잡도가 증가하는 것은 아닌지
2. Enum은 JSON형식으로 변환될 때  String이 되는 것 같은데, 이로 인해 iOS분들에게 괜한 혼란을 야기하는 것은 아닌지
이 부분이 고민이 되어 결국 boolean값으로 그대로 사용했습니다.
(대신 코드 중복 제거 문제는 메서드 분리로 해결했습니다.)
혹시 Enum과 boolean 중 어떤 자료형을 사용하는 게 더 바람직한지에 대해 괜찮으시다면 조언 부탁드립니다🙇‍♀️

미션 진행이 빠르지는 않지만, 지금까지 구현된 부분이라도 확인을 받고 싶어 조금 빠른 PR을 보냅니다.
부족한 코드지만 많은 피드백 부탁드립니다.
감사합니다! 😀

bibi6666667 committed:
initial commit

bibi6666667 committed:
Feat : 스프링 부트 파일 생성, index.html 테스트
스프링 부트 파일 생성 및 기본 기능 테스트

1song2 committed:
Initial commit

1song2 committed:
Remove unused code

1song2 committed:
Add class 'Card'

1song2 committed:
Add class 'Cards' and 'List'

1song2 committed:
Add class 'Activity'

1song2 committed:
Add class 'ActivityFeed'

1song2 committed:
Add class 'JSONParser' to parse JSON

1song2 committed:
Add class 'NetworkManager'

1song2 committed:
Create groups to organize files in

Jiwon-JJW committed:
Table 제작에 대한 내용 추가

Jiwon-JJW committed:
Merge pull request #1 from To-Do-List-Codesquad-Team04/bibiSpringBoot

기본 세팅에 대한 Merge

Jiwon-JJW committed:
Entity Card 추가

Jiwon-JJW committed:
테이블과 URL 정리를 위한 README 정리

Jiwon-JJW committed:
생성자 및 테이블 상태 표시하는 boolean 변수 추가

Jiwon-JJW committed:
Update README.md

1song2 committed:
Add title label and hamburger button

1song2 committed:
Rename 'ViewController' to 'BoardViewController'

Jiwon-JJW committed:
기본 getter, setter 추가 및 update, changeState 메서드 추가

bibi6666667 committed:
Feat : Card엔티티, CardController 생성

- Todo리스트에 들어갈 Card 엔티티를 생성
- Card 를 관리할 CardController 생성 및 기본 메서드 작성
- HomeController의 리턴값 변경

bibi6666667 committed:
Merge branch 'bibiSpringBoot' of https://github.com/To-Do-List-Codesquad-Team04/todo-list into bibiSpringBoot

bibi6666667 committed:
Feat : 테스트용 초기 데이터 삽입

bibi6666667 committed:
Feat : 테스트 데이터 조회를 위한 CardController 메서드 수정

bibi6666667 committed:
Feat : 테스트 데이터 조회를 위한 CardRepository 메서드 생성

Todo, Doing, Done 중 하나가 true면 Card를 List로 반환하는 메서드 생성

bibi6666667 committed:
Feat : 테스트를 위한 html form과 toString() 생성

bibi6666667 committed:
Feat : Todo 생성시 처리하는 newCard() 메소드 생성

Jiwon-JJW committed:
Merge branch 'Team-04' of https://github.com/To-Do-List-Codesquad-Team04/todo-list into Team-04

bibi6666667 committed:
Feat : API URL convention 수정

변경된 API URL을 README.md 에 반영

bibi6666667 committed:
Feat : Card의 상태를 변경하는 Move메서드 구현

Card를 옮길 때 Todo / Doing / Done 상태로 전환하는 세 메서드 구현

1song2 committed:
Add custom UIView 'ListHeaderView'

1song2 committed:
Embed 3 container views in 'BoardViewController' for card lists

bibi6666667 committed:
Feat : 불필요한 리턴값 삭제

포스트맨으로 테스트하기 위해 불필요한 리턴값 삭제

bibi6666667 committed:
Feat : 불필요한 리턴값 삭제

포스트맨으로 테스트하기 위해 불필요한 리턴값 삭제

bibi6666667 committed:
Fix : CardRepository가 CrudRepository를 확장하도록 수정

기존 JpaRepository를 확장하던 것에서 CrudRepository를 확장하도록 수정

bibi6666667 committed:
Fix : CardRepository가 CrudRepository를 확장하도록 수정

기존 JpaRepository를 확장하던 것에서 CrudRepository를 확장하도록 수정

1song2 committed:
Add custom UITableViewCell 'CardCell'

1song2 committed:
Make lists show cards

1song2 committed:
Change font size and font weight

1song2 committed:
Modify constraint

Jiwon-JJW committed:
Delete 및 Edit 메서드 수정 작업 진행

Jiwon-JJW committed:
Date Format을 지정하는 Annotation 추가

bibi6666667 committed:
README.md에 팀원 구성, 브랜치 관리 추가

bibi6666667 committed:
Feat : 테스트를 위한 index.html 수정

테스트를 위해 index.html에 카드 목록이 보이도록 코드 추가

bibi6666667 committed:
Feat : createDate -> dateTime으로 변수명 수정

bibi6666667 committed:
Feat : HTTP put, delete method를 위한 속성 추가

bibi6666667 committed:
Feat : dateTimeFormat 추가

bibi6666667 committed:
Feat : CardController 로직 수정 및 불필요한 리턴값 삭제

카드 수정, 삭제가 id값 기반으로 진행되도록 로직 수정
카드 추가 메서드에 불필요한 리턴값 삭제

1song2 committed:
Add space between cells

1song2 committed:
Add struct 'TodoData' to decode JSON

1song2 committed:
Modify 'JSONParser' and 'NetworkManager' according to 'TodoData' model

bibi6666667 committed:
Feat : Card에 불필요한 setter 삭제 및 delete() 추가

soft delete 방식의 delete() 추가

bibi6666667 committed:
Feat : CardService로 CardController 로직 분리

1song2 committed:
Assign segue identifiers

1song2 committed:
Comment 'cardBackgroundView'-related code

1song2 committed:
Modify constraint

1song2 committed:
Add stored property 'category' to class 'Card'

Change return type of JSON parsing method to optional array of 'Card'.

1song2 committed:
Add protocol 'ListViewControllerProtocol'

1song2 committed:
Get data using mockup API and update tableViews

Jiwon-JJW committed:
deleted 인스턴스 변수 추가 및 사용하지 않는 Annotation 제거

Jiwon-JJW committed:
welcome 페이지의 return값 변경

Jiwon-JJW committed:
delete 메서드 db반영을 위해 Repository에 save하도록 메서드 수정

Jiwon-JJW committed:
README 컨플릭트 해결

Jiwon-JJW committed:
RequestBody에서 데이터를 매칭해서 받도록 수정

Jiwon-JJW committed:
백엔드 폴더를 별도로 제작하여 프로젝트 파일 이동

Jiwon-JJW committed:
README 저장 경로 이동

bibi6666667 committed:
Feat : 데모를 위한 README.md 수정

작성한 구글 스프레드시트 링크 및 팀 규칙 추가

bibi6666667 committed:
Feat : BE 폴더 생성

bibi6666667 committed:
Feat : 기존 파일들을 BE디렉토리로 이동

wheejuni commented on BE/src/main/java/codesquad/TodoList/TodoListApplication.java:
```diff
@@ -0,0 +1,13 @@
+public class TodoListApplication {
+
+	public static void main(String[] args) {
+		SpringApplication.run(TodoListApplication.class, args);
```
Tab 설정이 4 spaces 같군요. 특별한 목적이 없다면 4 spaces가 더 좋진 않을까 생각해봅니다.

wheejuni commented on BE/src/main/java/codesquad/TodoList/domain/Card.java:
```diff
@@ -0,0 +1,111 @@
+    @JsonFormat(pattern = "yyyy-MM-dd hh:mm:ss", timezone = "Asia/Seoul")
+    private LocalDateTime dateTime = LocalDateTime.now();
+
+    private boolean todo;
+    private boolean doing;
+    private boolean done;
+    private boolean deleted;
```
절대로 좋은 구현이 아닙니다. 이렇게 할 바에야 `enum` 을 쓰는게 낫습니다.
클라이언트 개발자들도 그 편을 선호하지 않을까요? 이 카드의 상태를 특정하기 위해 네 개의 필드를 참조하는 것보다 사전에 합의된 `enum` 을 사용하는 편이 어떻게 봐도 더 좋은 구현일 것 같습니다.

bibi6666667 replied:
아 그렇군요! todo, doing, done, deleted 네 상태를 Status라는 enum클래스를 사용하도록 수정하겠습니다. 감사합니다😀👍!

wheejuni commented on BE/src/main/java/codesquad/TodoList/service/CardService.java:
```diff
@@ -0,0 +1,66 @@
+        this.cardRepository = cardRepository;
+    }
+
+    public void createTodo(Card card) {
+        card.changeStateTodo();
+        cardRepository.save(card);
+    }
+
+    public void createDoing(Card card) {
+        card.changeStateDoing();
+        cardRepository.save(card);
+    }
+
+    public void createDone(Card card) {
+        card.changeStateDone();
+        cardRepository.save(card);
+    }
```
`enum` 설계해서 개선되면 이 메소드들도 다 걷어낼 수 있겠네요.

bibi6666667 replied:
넵 enum 사용해서 이 메서드들은 삭제하겠습니다!

wheejuni reviewed (changes requested):
초기 세팅을 잡느라 고생이 많으셨을 것 같네요. 수고 많으셨습니다. 👍 
우선 이 경우엔 `enum` 을 사용하는 것이 맞는 방향으로 생각됩니다. 
네 개라는 경우의 수는 결코 적지 않고요, 특히 상태 하나를 만들기 위해 나머지 세 개 변수의 상태를 바꿔야 하는 것은 비효율적인 것으로 생각됩니다. 

수고 많으셨습니다. 💯

Jiwon-JJW committed:
boolean이었던 status들을 enum으로 변경하여 메소드 변경 및 수정 작업 진행

bibi6666667 committed:
Merge branch 'bibiSpringBoot' of https://github.com/To-Do-List-Codesquad-Team04/todo-list into bibiSpringBoot

bibi6666667 committed:
Feat : Card의 내부클래스로 Status enum 추가

Card의 상태를 용이하게 나타내기 위해 기존 boolean이었던 Todo, Doing, Done, Deleted 값을 내부클래스 enum으로 변경
CardRepository에 새로운 메서드를 정의해 CardService에서 사용

bibi6666667 committed:
Feat : 초기데이터의 Deleted 삭제 및 status 변경

Deleted가 Status에 포함됨에 따라 Deleted항을 삭제하고 status값을 변경

1song2 committed:
Add struct 'MockJSON'

1song2 committed:
Add section headers to tableView for spacing

Jiwon-JJW committed:
Merge pull request #2 from To-Do-List-Codesquad-Team04/bibiSpringBoot

enum으로 변경 및 메서드 수정

Jiwon-JJW commented:
안녕하세요! 
boolean 으로 표시되었던 status를 enum으로 변경 하여 재 PR요청을 드립니다!
다만, 작성하여 주신 리뷰에서 Tab 설정에 대한 말씀이 이해가 잘 되지 않는데, 혹시 다시 한 번 설명 부탁드려도 될까요?😅
리뷰 감사드립니다!

1song2 committed:
Change tableView style to inset grouped

1song2 committed:
Add trailing swipe action to delete card

1song2 committed:
Add popupView to add card

1song2 committed:
Merge three containerViewController into one

1song2 committed:
Remove unused code

1song2 committed:
Separate UITableViewDataSource

1song2 committed:
Rename variable to 'notes'

1song2 committed:
Add protocol for PopupView to create card to list

1song2 committed:
Add method to display and update count of cards

1song2 committed:
Fix error in trailing swipe action

1song2 committed:
Let Card conform to protocols

1song2 committed:
Add class 'CacheDragCoordinator' for drag session

1song2 committed:
Let 'ListViewController' conform to UITableViewDragDelegate to drag card

1song2 committed:
Let 'ListViewController' conform to UITableViewDropDelegate to drop card

1song2 committed:
Merge branch 'iOS/data-model' into Team-04

1song2 committed:
Merge branch 'iOS/feature/layout' into Team-04

1song2 committed:
Merge branch 'iOS/feature/drag-and-drop' into Team-04

1song2 committed:
Create iOS folder to organize files in
